File: predict_noxn.py
"""Use random forests to predict nitrate concentrations globally."""
import os
import tempfile
import shutil

from osgeo import gdal
import numpy
import pandas

from sklearn.ensemble import RandomForestRegressor
from joblib import dump, load
import pygeoprocessing

# full dict of covariate datasets at native resolution
_GLOBAL_COVARIATE_PATH_DICT = {
    'average_flow': "F:/NCI_NDR/Data streamflow FLO1K/average_flow_1990_2015.tif",
    'flash_flow': "F:/NCI_NDR/Data streamflow FLO1K/mean_div_range_1990_2015.tif",
    'precip_variability': "F:/NCI_NDR/Data precip Worldclim/wc2.0_bio_5m_15.tif",
    'population': "F:/NCI_NDR/Data population HYDE3.2/inhabitants_avg_1990_2015.tif",
    'proportion_urban': "F:/NCI_NDR/Data urban extent GRUMP/perc_urban_5min.tif",
    'percent_no_sanitation': "F:/NCI_NDR/Data sanitation/no_sanitation_provision_avg_2000-2015.tif",
    'depth_to_groundwater': "F:/NCI_NDR/Data HydroATLAS/gwt_cm_sav_level12.tif",
    'clay_percent': "F:/NCI_NDR/Data HydroATLAS/cly_pc_sav_level12.tif",
    'silt_percent': "F:/NCI_NDR/Data HydroATLAS/slt_pc_sav_level12.tif",
}

# n export rasters for a set of scenarios: each must be aggregated up to
# approximately 5 arc min resolution, each must have a unique basename
_N_EXPORT_PATH_DICT = {
    'baseline': "F:/NCI_NDR/Data NDR/nutrient_deficit_5min_cur_compressed_md5_031d4bb444325835315a2cc825be3fd4.tif",
    # continue with other scenarios here
}

# directory to hold temporary outputs
_PROCESSING_DIR = "F:/NCI_NDR/rf_processing"

# noxn and covariate observations for groundwater
_NOXN_PREDICTOR_GR_DF_PATH = "C:/Users/ginge/Documents/Python/nci_ndr/noxn_predictor_df_gr.csv"

# noxn and covariate observations for surface water
_NOXN_PREDICTOR_SURF_DF_PATH = "C:/Users/ginge/Documents/Python/nci_ndr/noxn_predictor_df_surf.csv"

# nodata value for inputs and result
_TARGET_NODATA = -1.0

def reclassify_nodata(target_path):
    """Reclassify the nodata value of a raster to _TARGET_NODATA.

    Check to see if the current nodata value for the raster indicated by
    `target_path` is _TARGET_NODATA. If not, convert all areas of nodata in the
    target raster to _TARGET_NODATA and set the nodata value for the taret
    raster to _TARGET_NODATA.

    Parameters:
        target_path (string): path to target raster

    Side effects:
        modifies the raster indicated by `target_path`

    Returns:
        None

    """
    def reclassify_op(target_raster):
        reclassified_raster = numpy.copy(target_raster)
        reclassify_mask = (target_raster == previous_nodata_value)
        reclassified_raster[reclassify_mask] = _TARGET_NODATA
        return reclassified_raster

    previous_nodata_value = pygeoprocessing.get_raster_info(
        target_path)['nodata'][0]
    if numpy.isclose(previous_nodata_value, _TARGET_NODATA):
        return

    file_handle, temp_path = tempfile.mkstemp(dir=_PROCESSING_DIR)
    shutil.copyfile(target_path, temp_path)
    pygeoprocessing.raster_calculator(
        [(temp_path, 1)], reclassify_op, target_path, gdal.GDT_Float32,
        _TARGET_NODATA)

    # clean up
    os.close(file_handle)
    os.remove(temp_path)


def prepare_covariates(predictor_names, aligned_covariate_dir):
    """Align covariate rasters and ensure that they all share one nodata value.

    Parameters:
        predictor_names (list): list of predictor names giving the order of
            covariates used to fit the random forests model.
        aligned_covariate_dir (string): path to directory where aligned
            covariate rasters should be stored

    Side effects:
        creates or modifies the directory `aligned_covariate_dir`
        creates or modifies rasters in `aligned_covariate_dir`, one for each
            predictor variable in `predictor_names`

    Returns:
        a dictionary whose keys are the covariate predictors in
            `predictor_names` and whose values contain paths to aligned
            covariate rasters

    """
    if not os.path.exists(aligned_covariate_dir):
        os.makedirs(aligned_covariate_dir)
    target_pixel_size = pygeoprocessing.get_raster_info(
        _GLOBAL_COVARIATE_PATH_DICT['average_flow'])['pixel_size']

    input_path_list = [
        _GLOBAL_COVARIATE_PATH_DICT[covar] for covar in predictor_names]
    aligned_path_list = [
        os.path.join(aligned_covariate_dir, os.path.basename(
            _GLOBAL_COVARIATE_PATH_DICT[covar])) for covar in predictor_names]
    if not all([os.path.isfile(r) for r in aligned_path_list]):
        pygeoprocessing.align_and_resize_raster_stack(
            input_path_list, aligned_path_list,
            ['near'] * len(input_path_list),
            target_pixel_size, 'intersection')
    # set nodata value for all aligned covariates to _TARGET_NODATA
    for path in aligned_path_list:
        reclassify_nodata(path)
    aligned_covariate_dict = {
        covar: os.path.join(aligned_covariate_dir, os.path.basename(
            _GLOBAL_COVARIATE_PATH_DICT[covar]))
        for covar in predictor_names
    }
    return aligned_covariate_dict


def predict_noxn(
        aligned_covariate_dir, n_export_path, rf_pickle_filename,
        predictor_names, output_path):
    """Generate global nitrate concentrations from a set of predictors.

    Parameters:
        aligned_covariate_dir (string): path to directory where aligned
            covariate rasters should be stored
        n_export_path (string): path to raster containing N export for a single
            scenario
        rf_pickle_filename (string): path to file on disk containing trained
            random forests model. This file should have the extension '.joblib'
        predictor_names (list): list of predictor names giving the order of
            covariates used to fit the random forests model. Any subsequent use
            of the model to make predictions on new covariate data must use
            covariates in this order.
        output_path (string): path to location on disk where output should be
            saved. This output will be a tif containing predicted nitrate
            concentration in mg/L

    Side effects:
        creates a tif giving predicted nitrate concentration in mg/L at the
            location given by `output_path`

    """
    def rf_predict_op(*covariate_list):
        """Use a trained random forest model to predict noxn."""
        invalid_mask = numpy.any(
            numpy.isclose(numpy.array(covariate_list), _TARGET_NODATA), axis=0)
        covar_arr = numpy.stack([r.ravel() for r in covariate_list], axis=1)
        noxn_arr = rf_model.predict(covar_arr)
        result = noxn_arr.reshape(covariate_list[0].shape)
        result[invalid_mask] = _TARGET_NODATA
        return result

    _GLOBAL_COVARIATE_PATH_DICT['n_export'] = n_export_path
    aligned_covariate_dict = prepare_covariates(
        predictor_names, aligned_covariate_dir)
    covariate_path_list = [
        aligned_covariate_dict[key] for key in predictor_names]

    rf_model = load(rf_pickle_filename)
    pygeoprocessing.raster_calculator(
        [(path, 1) for path in covariate_path_list],
        rf_predict_op, output_path, gdal.GDT_Float32, _TARGET_NODATA)


def train_rf_model(
        noxn_predictor_df_path, max_features, rf_pickle_filename):
    """Train random forest model using observed data.

    Parameters:
        noxn_predictor_df_path (string): path to csv containing noxn
            obserations and predictor values for monitoring points. Must
            contain a column named 'noxn', containing observed nitrate
            concentrations, which is the target of the random forests model
        max_features (int): maximum number of variables to use in each tree
        rf_pickle_filename (string): location on disk where trained random
            forests model should be saved. This file should have the extension
            'joblib'

    Side effects:
        saves an instance of sklearn.ensemble.RandomForestRegressor, which has
            been trained on the observed data provided, and may be used to make
            predictions on new data, at the location `rf_pickle_filename`

    Returns:
        a list of predictor names giving the order of covariates used to fit
            the model. Any subsequent use of the model to make predictions on
            new covariate data must use covariates in this order.

    """
     # read data that was filtered and subsetted in R
    combined_df = pandas.read_csv(noxn_predictor_df_path)
    # drop rows containing missing data
    combined_df.dropna(inplace=True)

    # Separate data frame into response and predictors
    noxn_arr = numpy.array(combined_df['noxn'])
    predictor_df = combined_df.drop('noxn', axis=1)
    predictor_arr = numpy.array(predictor_df)

    # train random forests model
    # Instantiate model, using parameters chosen to match those used in R
    rf_model = RandomForestRegressor(
        random_state=42, n_estimators=500, criterion="mse",
        max_features=max_features, min_samples_leaf=5, oob_score=True,
        bootstrap=True)
    rf_model.fit(predictor_arr, noxn_arr)
    dump(rf_model, rf_pickle_filename)

    predictor_names = list(predictor_df.columns)
    return predictor_names


def surface_noxn_workflow():
    """Workflow to predict noxn in surface water."""
    surface_rf_path = os.path.join(_PROCESSING_DIR, 'surface_model.joblib')
    aligned_covariate_dir = os.path.join(
        _PROCESSING_DIR, 'aligned_covariates_surface')
    surface_max_features = 7
    surface_predictors = train_rf_model(
        _NOXN_PREDICTOR_SURF_DF_PATH, surface_max_features, surface_rf_path)
    for scenario_key in _N_EXPORT_PATH_DICT:
        n_export_path = _N_EXPORT_PATH_DICT[scenario_key]
        output_path = os.path.join(
            _PROCESSING_DIR, 'surface_noxn_{}.tif'.format(scenario_key))
        predict_noxn(
            aligned_covariate_dir, n_export_path, surface_rf_path,
            surface_predictors, output_path)


def groundwater_noxn_workflow():
    """Workflow to predict noxn in groundwater."""
    ground_rf_path = os.path.join(_PROCESSING_DIR, 'ground_model.joblib')
    aligned_covariate_dir = os.path.join(
        _PROCESSING_DIR, 'aligned_covariates_ground')
    ground_max_features = 2
    ground_predictors = train_rf_model(
        _NOXN_PREDICTOR_GR_DF_PATH, ground_max_features, ground_rf_path)
    for scenario_key in _N_EXPORT_PATH_DICT:
        n_export_path = _N_EXPORT_PATH_DICT[scenario_key]
        output_path = os.path.join(
            _PROCESSING_DIR, 'ground_noxn_{}.tif'.format(scenario_key))
        predict_noxn(
            aligned_covariate_dir, n_export_path, ground_rf_path,
            ground_predictors, output_path)


def main():
    """Program entry point."""
    surface_noxn_workflow()
    groundwater_noxn_workflow()


if __name__ == '__main__':
    __spec__ = None
    main()

File: predict_noxn.R
# Make predictions from random forest
library(ranger)
library(raster)

# extent bounding boxes of a regular grid of 60 tiles covering the extent of covariates
EXTENT_DF <- read.csv("C:/Users/ginge/Documents/Python/nci_ndr/extent_df.csv")

# aligned covariate rasters
COVAR_PATH_LIST <- list(
  'precip_variability'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/wc2.0_bio_5m_15.tif",
  'population'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/inhabitants_avg_1990_2015.tif",
  'pigs'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/5_Pg_2010_Da.tif",
  'cattle'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/5_Ct_2010_Da.tif",
  'flash_flow'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/mean_div_range_1990_2015.tif",
  'average_flow'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/average_flow_1990_2015.tif",
  'percent_no_sanitation'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/no_sanitation_provision_avg_2000-2015.tif",
  'proportion_urban'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/perc_urban_5min.tif",
  'sand_percent'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/SNDPPT_M_sl1_10km_ll.tif",
  'clay_percent'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/CLYPPT_M_sl1_10km_ll.tif",
  'depth_to_groundwater'="C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/gwt_cm_sav_level12.tif"
  )

# aligned n export rasters for a set of scenarios
N_EXPORT_PATH_LIST = list(
  'extensification_bmps_irrigated'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_extensification_bmps_irrigated_300.0_D8_export_reduce28sum_md5_f28fd3b8e444044165bf8356f16da2cb.tif',
  'extensification_bmps_rainfed'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_extensification_bmps_rainfed_300.0_D8_export_reduce28sum_md5_6bf450d1e48ce523a96ef39971d14dff.tif',
  'extensification_current_practices'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_extensification_current_practices_300.0_D8_export_reduce28sum_md5_a7f97c7ea643732dcb9a37a38261d9be.tif',
  'extensification_intensified_irrigated'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_extensification_intensified_irrigated_300.0_D8_export_reduce28sum_md5_91b8c02c962a3ded85d7437a04544518.tif',
  'extensification_intensified_rainfed'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_extensification_intensified_rainfed_300.0_D8_export_reduce28sum_md5_1160ef8f4fb9b51a92c45fa7c081da97.tif',
  'fixedarea_bmps_irrigated'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_fixedarea_bmps_irrigated_300.0_D8_export_reduce28sum_md5_ce57595bd7f1aa0ee0b5aba91c5e5417.tif',
  'fixedarea_bmps_rainfed'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_fixedarea_bmps_rainfed_300.0_D8_export_reduce28sum_md5_b7d9eda12a76ceb6c5ed72501304318f.tif',
  'fixedarea_intensified_irrigated'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_fixedarea_intensified_irrigated_300.0_D8_export_reduce28sum_md5_5ed83d7dd073df9dd2ff5e55f493d1c9.tif',
  'fixedarea_intensified_rainfed'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_fixedarea_intensified_rainfed_300.0_D8_export_reduce28sum_md5_41c8043f4399419fcc995117be7c2a8f.tif',
  'grazing_expansion'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_grazing_expansion_300.0_D8_export_reduce28sum_md5_ebc5c0c91bbabff356fa1a0c57bfc01c.tif',
  'restoration'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_restoration_300.0_D8_export_reduce28sum_md5_0f7080ca2c1cf5b8ee10d7b39e3cce85.tif',
  'sustainable_currentpractices'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/compressed_sustainable_currentpractices_300.0_D8_export_reduce28sum_md5_11cbca90d8c151b07336aba2e917ec05.tif',
  'baseline_currentpractices'='C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_5.15.20/subset_2000_2015/intermediate/aligned_covariates_ground/baseline_currentpractices_300.0_D8_export.tif'
)

# Use random foresets model to predict se from a raster stack
predict_se <- function(covar_stack, model, save_as) {
  out <- raster(covar_stack, layer=0)
  bs <- blockSize(out)
  out <- writeStart(out, save_as, overwrite=TRUE)
  for (i in 1:bs$n) {
    covar_df <- as.data.frame(getValues(covar_stack, row=bs$row[i], nrows=bs$nrows[i]))
    na_vals <- apply(covar_df, 1, function(x) sum(is.na(x)))
    p <- numeric(length=nrow(covar_df))
    p[na_vals > 0] <- NA
    if (length(p[na_vals==0]) > 0) {
      p[na_vals==0] <- raster::predict(
        model, covar_df[na_vals==0, ], type='se')$se
    }
    out <- writeValues(out, p, bs$row[i])
  }
  out <- writeStop(out)
  return(out)
}

# Predict response from a raster stack
predict_response <- function(covar_stack, model, save_as) {
  out <- raster(covar_stack, layer=0)
  bs <- blockSize(out)
  out <- writeStart(out, save_as, overwrite=TRUE)
  for (i in 1:bs$n) {
    covar_df <- as.data.frame(getValues(covar_stack, row=bs$row[i], nrows=bs$nrows[i]))
    na_vals <- apply(covar_df, 1, function(x) sum(is.na(x)))
    p <- numeric(length=nrow(covar_df))
    p[na_vals > 0] <- NA
    if (length(p[na_vals==0]) > 0) {
      p[na_vals==0] <- raster::predict(
        model, covar_df[na_vals==0, ], type='response')$predictions
    }
    out <- writeValues(out, p, bs$row[i])
  }
  out <- writeStop(out)
  return(out)
}

# Make response and se predictions by tiling
# Parameters:
#   - intermediate_dir (string): directory where intermediate tiled results
#     should be writting
#   - EXTENT_DF (data frame): data frame containing bounding boxes for tiles
#     covering the globe
#   - covariate_name_list (named list): named list giving covariates. The order
#     of covariates in this list must match the order of predictors that were
#     used to train the random forests model
#   - covariate_path_list (named list): list indexed by covariate name and
#     giving paths to aligned covariate rasters for the globe
#   - ranger_model (model object): trained ranger model that should be used to
#     make predictions
#   - response_target_path (string): path to location on disk where noxn
#     global raster should be saved
#   - se_target_path (string): path to location on disk where standard error
#     global raster should be saved
tile_and_predict <- function(
    intermediate_dir, EXTENT_DF, covariate_name_list, covariate_path_list,
    ranger_model, response_target_path, se_target_path) {
  native_format_list <- list()
  covar_tile_dir <- paste(intermediate_dir, 'covar_tile', sep='/')
  dir.create(covar_tile_dir, showWarnings=FALSE)
  pred_tile_dir <- paste(intermediate_dir, 'pred_tile', sep='/')
  dir.create(pred_tile_dir, showWarnings=FALSE)
  merge_list_se <- list()
  merge_list_response <- list()
  for (r in 1:NROW(EXTENT_DF)) {
    extent <- extent(as.numeric(EXTENT_DF[r, ]))
    for(cv in covariate_name_list){
      save_as <- paste(covar_tile_dir, paste(
        tools::file_path_sans_ext(
          basename(covariate_path_list[[cv]])), '.grd', sep=''),
        sep='/')
      crop(raster(covariate_path_list[[cv]]), y=extent,
           filename=save_as, overwrite=TRUE)
      native_format_list[[cv]] <- save_as
    }
    covar_stack <- stack(lapply(native_format_list, FUN=raster))
    names(covar_stack) <- covariate_name_list

    # predict response
    response_pred_tile_filename <- paste(
      pred_tile_dir, paste("response_", r, ".tif", sep=''), sep='/')
    response_pred <- predict_response(
      covar_stack, model=ranger_model, save_as=response_pred_tile_filename)
    merge_list_response[[r]] <- response_pred_tile_filename

    # predict standard error of the response
    se_pred_tile_filename <- paste(
      pred_tile_dir, paste("se_pred_", r, ".tif", sep=''), sep='/')
    se_pred <- predict_se(
      covar_stack, model=ranger_model, save_as=se_pred_tile_filename)
    merge_list_se[[r]] <- se_pred_tile_filename
  }
  # merge the tiles together
  merged_response <- Reduce(merge, lapply(merge_list_response, raster))
  writeRaster(merged_response, filename=response_target_path)
  merged_se <- Reduce(merge, lapply(merge_list_se, raster))
  writeRaster(merged_se, filename=se_target_path)

  # clean up
  unlink(covar_tile_dir, recursive=TRUE)
  unlink(pred_tile_dir, recursive=TRUE)
}

# surface noxn
# noxn and covariate observations for surface water
NOXN_PREDICTOR_SURF_DF_PATH = "C:/Users/ginge/Documents/Python/nci_ndr/noxn_predictor_df_surf_2000_2015.csv"

# train the random forest model
train_df = read.csv(NOXN_PREDICTOR_SURF_DF_PATH)
ranger_train_df <- train_df[complete.cases(train_df), ]
ranger_model <- ranger(noxn~., keep.inbag=TRUE, data=ranger_train_df,
                       mtry=2, min.node.size=5, importance='impurity')  # using mtry and min.node.size tuned by caret

output_dir <- "C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_3.27.21/R_ranger_pred"
dir.create(output_dir, recursive=TRUE, showWarnings=FALSE)
intermediate_dir <- "C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_3.27.21/R_ranger_pred/surf_intermediate"
dir.create(intermediate_dir)
covariate_df <- subset(train_df, select=-c(noxn))
covariate_name_list <- colnames(covariate_df)
for (scenario in names(N_EXPORT_PATH_LIST)) {
  COVAR_PATH_LIST[['n_export']] = N_EXPORT_PATH_LIST[[scenario]]
  response_target_path <- paste(output_dir, paste(
    "surface_noxn_", scenario, ".tif", sep=''), sep='/')
  se_target_path <- paste(output_dir, paste(
    "surface_noxn_se_", scenario, ".tif", sep=''), sep='/')
  tile_and_predict(
    intermediate_dir, EXTENT_DF, covariate_name_list, COVAR_PATH_LIST,
    ranger_model, response_target_path, se_target_path)
}

# ground noxn
# GEMStat, China, USGS, and Ouedraogo (Africa) combined
NOXN_PREDICTOR_GR_DF_PATH = "C:/Users/ginge/Dropbox/NatCap_backup/NCI WB/noxn_predictor_df_gr_Ouedraogo_USGS_China_GEMStat.csv"

# train the random forest model
train_df <- read.csv(NOXN_PREDICTOR_GR_DF_PATH)
ranger_train_df <- train_df[complete.cases(train_df), ]
ranger_ground_model <- ranger(
  noxn~., keep.inbag=TRUE, data=ranger_train_df, mtry=2, min.node.size=5,
  importance='impurity')  # using mtry and min.node.size tuned by caret

output_dir <- "C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_3.27.21/Ouedraogo_USGS_China_GEMStat/R_ranger_pred"
dir.create(output_dir, recursive=TRUE, showWarnings=FALSE)
intermediate_dir <- "C:/Users/ginge/Documents/NatCap/GIS_local/NCI_NDR/Results_3.27.21/Ouedraogo_USGS_China_GEMStat/R_ranger_pred/ground_intermediate"
dir.create(intermediate_dir, recursive=TRUE, showWarnings=FALSE)
covariate_df <- subset(train_df, select=-c(noxn))
covariate_name_list <- colnames(covariate_df)
for (scenario in names(N_EXPORT_PATH_LIST)) {  # c("fixedarea_currentpractices")) {
  COVAR_PATH_LIST[['n_export']] = N_EXPORT_PATH_LIST[[scenario]]
  response_target_path <- paste(output_dir, paste(
    "ground_noxn_", scenario, ".tif", sep=''), sep='/')
  se_target_path <- paste(output_dir, paste(
    "ground_noxn_se_", scenario, ".tif", sep=''), sep='/')
  if(!file.exists(response_target_path)){
    tile_and_predict(
      intermediate_dir, EXTENT_DF, covariate_name_list, COVAR_PATH_LIST,
      ranger_ground_model, response_target_path, se_target_path)
  }
}
